### ISSUE TITLE
Add a way to quantify plan progress

### ISSUE THREAD

**Issue by stan-dot** [DiamondLightSource/blueapi#605]
As a user I would like to see the progress of the plan I defined, be it in points scanned out of total number, or some other metric.

**Comment by callumforrester** [comment]
`ProgressEvents` provide a path towards this, but they are currently per `Status` object i.e. per motor-move, detector trigger etc.

Obviously there is no universal metric for the progress of a plan, it depends on what the plan is trying to accomplish. Some plans cannot quantify their progress very well at all. For example an adaptive scan may have no idea of its total number of points upfront.

For this reason I think it is best to have an optional way to include a self-reporting mechanism in a plan itself, maybe something similar to TQDM. For example:

```python
def my_scan(points: Spec[Movable]) -> MsgGenerator:
    motors = points.axes()
    yield from bps.mv(motors, starting_point)
    for point in plan_progress(points.frames()):
        yield from bps.mv(motors)
```

`plan_progress` would insert a side effect updating a status that the `RunEngine` knows about.

It is then the responsibility of a plan author to define how to report progress. How does that sound to you @stan-dot?

Also, would this then become an issue on bluesky itself rather than blueapi?

**Comment by stan-dot** [comment]
to have a default, overridable way which for all scans in the form `start stop step` can be just the number of points. Given that that is the most run functionality currently, adding a plan progress bar just for this simple use case will make a big difference to the end user

**Comment by stan-dot** [comment]
the length of the plan would be estimated from the params.

**Comment by callumforrester** [comment]
So you want to inspect the params for `start`, `stop` and `num`, say? But then how/where does the default code that detects what `num` is and how much progress the scan has made towards `stop` actually go? Inside the `RunEngine`?

**Comment by stan-dot** [comment]
that's an implementation thing, the _issue_ is about a user need. 

One implementation could be handling the start an stop in the client side and listen for 'step' events, and increment `count`

**Comment by callumforrester** [comment]
This discussion is focused on implementation, as the requirements have already been agreed upon. I believe this calculation should be handled server-side, without relying on a default that may or may not suit the specific plan. Implementing it as suggested could lead to the following issues:
- It must be reimplemented many times, once per client
- It has no way to know if it misses a step event and gets out of sync
- It will behave in a very confusing way for adaptive scans, while I agree we should cover the 80% case, let's leave the 20% case in a neutral state rather than make it worse
- It thickens the client, we want a thinner client than GDA with as much business logic on the server side as possible.

**Comment by stan-dot** [comment]
of course the adaptive scans won't have this

re: missing a step - I thought that the RMQ message bus has a delivery guarantee. 

> But then how/where does the default code that detects what num is and how much progress the scan has made towards stop actually go? Inside the RunEngine?

I guess the server-side implementation would likely use some `callback` logic, but I might not be the best person to dream up the details here

**Comment by callumforrester** [comment]
Fair enough, if we agree on server-side I think the decision I want out of this discussion is whether the implementation lives in blueapi or actual core bluesky (in which case I will close this issue and raise one there).

I think @coretl will advocate for the latter since he is keen on the workflow of developing plans outside of blueapi and running them for users inside it.

**Comment by stan-dot** [comment]
@coretl 's description is a more verbose version  of what I had in mind when writing
> One implementation could be handling the start an stop in the client side and listen for 'step' events, and increment count

@callumforrester 

This aligns with the websockets stuff, and the flow of information the other way, or with the bluesky-stomp library as used in blueapi. we could add an endpoint

```python
def get_plan_progress(job_id: string): Tuple[number, number]:
    ```

**Comment by graeme-winter** [comment]
Rather than trying to work out a general way to measure progress: the author of the plan should be able to publish the relative progress even if it is not linear, so could we provide a method (like `tqdm`) which allows the plan author to provide updates from time to time, and a general API which can be used to transmit these to the end user? Similar idea to logging.

There are other progress bar implementations which allow periodic updates, which could also be used as a model. This does not need to be perfect, and for adaptive scans it could also reasonably go retrograde if the plan suddenly decides that it wants to re-scan a region at a higher resolution.

I assume that this is really a courtesy to the end user to see how long it will _probably_ be until this scan is complete right?
